sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateSafeProjection.scala

```diff
@@ -0,0 +1,148 @@
+
+  protected def bind(in: Seq[Expression], inputSchema: Seq[Attribute]): Seq[Expression] =
+    in.map(BindReferences.bindReference(_, inputSchema))
+
+  private def genUpdater(
+    ctx: CodeGenContext,
+    setter: String,
+    dataType: DataType,
+    ordinal: Int,
+    value: String): String = {
+    dataType match {
+      case StringType => s"$setter.update($ordinal, $value.clone())"
+      case struct: StructType =>
+        val rowTerm = ctx.freshName("row")
+        val updates = struct.map(_.dataType).zipWithIndex.map { case (dt, i) =>
+          val colTerm = ctx.freshName("col")
+          s"""
+             if ($value.isNullAt($i)) {
+               $rowTerm.setNullAt($i);
+             } else {
+               ${ctx.javaType(dt)} $colTerm = ${ctx.getValue(value, dt, s"$i")};
+               ${genUpdater(ctx, rowTerm, dt, i, colTerm)};
+             }
+           """
+        }.mkString("\n")
+        s"""
+           $genericMutableRowType $rowTerm = new $genericMutableRowType(${struct.fields.length});
+           $updates
+           $setter.update($ordinal, $rowTerm.copy());
+         """
+      case _ =>
+        ctx.setColumn(setter, dataType, ordinal, value)
+    }
+  }
+
+  protected def create(expressions: Seq[Expression]): Projection = {
+    val ctx = newCodeGenContext()
+    val projectionCode = expressions.zipWithIndex.map {
+      case (NoOp, _) => ""
+      case (e, i) =>
+        val evaluationCode = e.gen(ctx)
+        evaluationCode.code +
+          s"""
+            if (${evaluationCode.isNull}) {
+              mutableRow.setNullAt($i);
+            } else {
+              ${genUpdater(ctx, "mutableRow", e.dataType, i, evaluationCode.primitive)};
+            }
+          """
+    }
+    // collect projections into blocks as function has 64kb codesize limit in JVM
+    val projectionBlocks = new ArrayBuffer[String]()
+    val blockBuilder = new StringBuilder()
+    for (projection <- projectionCode) {
+      if (blockBuilder.length > 16 * 1000) {
+        projectionBlocks.append(blockBuilder.toString())
+        blockBuilder.clear()
+      }
+      blockBuilder.append(projection)
+    }
+    projectionBlocks.append(blockBuilder.toString())
+
+    val (projectionFuns, projectionCalls) = {
+      // inline it if we have only one block
+      if (projectionBlocks.length == 1) {
+        ("", projectionBlocks.head)
+      } else {
+        (
+          projectionBlocks.zipWithIndex.map { case (body, i) =>
+            s"""
+               |private void apply$i(InternalRow i) {
+               |  $body
+               |}
+             """.stripMargin
+          }.mkString,
+          projectionBlocks.indices.map(i => s"apply$i(i);").mkString("\n")
+          )
+      }
+    }
+
+    val code = s"""
+      public Object generate($exprType[] expr) {
+        return new SpecificProjection(expr);
+      }
+
+      class SpecificProjection extends ${classOf[BaseProjection].getName} {
```

Comment:
change it to a different name; i noticed it'd help debugging because this classname does show up in exception stacktraces.